### PR TITLE
[`pylint`] Implement `import-private-name` (`C2701`)

### PR DESCRIPTION
**Pull request by tjkuson**
## Summary

Implements  [`import-private-name` (`C2701`)](https://pylint.pycqa.org/en/latest/user_guide/messages/convention/import-private-name.html) as  `import-private-name` (`PLC2701`). Includes documentation.

Related to #970.

Closes https://github.com/astral-sh/ruff/issues/9138.

### PEP 420 namespace package limitation

`checker.module_path` doesn't seem to support automatic detection of namespace packages (PEP 420). This leads to 'false' positives (Pylint allows both).

Currently, for this to work like Pylint, users would have to [manually input known namespace packages](https://beta.ruff.rs/docs/settings/#namespace-packages).

## Test Plan

`cargo test`

### CODE REVIEW

**Comment by tjkuson** [comment]
Most of the triggers from the ecosystem check seem to be in test suites, which is fine. The others seem to be false positives from not checking if a module is importing something from itself.

**Comment by tjkuson** [comment]
Looking through the ecosystem check, the flagged violations are all:

- true positives,
- issues with PEP 420 (which can be fixed using the [`namespace-packages` setting](https://beta.ruff.rs/docs/settings/#namespace-packages), or
- test suites importing private names for testing reasons (which makes sense)

**Comment by tjkuson** [comment]
Related: #6114.

File: crates/ruff/src/rules/pylint/rules/import_private_name.rs
Comment (konstin) [inline crates/ruff/src/rules/pylint/rules/import_private_name.rs:54]:
```suggestion
                format!("Private name import `{name}` from external module `{module}`")
```

**Review by konstin — APPROVED** [review]

File: crates/ruff/src/rules/pylint/rules/import_private_name.rs
Comment (tjkuson) [reply crates/ruff/src/rules/pylint/rules/import_private_name.rs:77]: This is what Pylint does, so I think you are right.

**Comment by AiyionPrime** [comment]
@tjkuson Is there anything I can do to help moving this forward?
I'd really like to have this checker in ruff.

Or is all that's left to to a rebase and the review of @charliermarsh?

**Comment by tjkuson** [comment]
> @tjkuson Is there anything I can do to help moving this forward? I'd really like to have this checker in ruff.
> 
> Or is all that's left to to a rebase and the review of @charliermarsh?

Wow, I forgot about this PR! If I recall correctly, this PR was working but required users to specify `namespace` packages (otherwise, it would false positive). I'm not a maintainer, but perhaps the team are waiting for #6114 to be solved before accepting this merge?

@charliermarsh I don't mind rebasing and tweaking the PR if you are open to it being merged.

**Comment by AiyionPrime** [comment]
If wanted I'd be open to test it prior merging.
Obviously I'm not a maintainer either, but to my understanding C2701 is working, but has a known limitation for projects with namespace packages, which are not all projects.

Might a note in the docs regarding this behavior be sufficient?
It feels like #6114 would be a nice addition but is a quite orthogonal feature?
Or are there likely changes necessary in this PR once it is merged?

Anyway, thanks for the fast response, let me know when I can help :)

**Comment by charliermarsh** [comment]
I think we probably _do_ need to support the type annotation thing, since we changed `flake8-self` to support this: https://github.com/astral-sh/ruff/pull/9036

**Comment by charliermarsh** [comment]
How does this compare to the flake8-self rules?

**Comment by tjkuson** [comment]
> I think we probably _do_ need to support the type annotation thing, since we changed `flake8-self` to support this: #9036

Sure! It should be a simple change if you are still open to the rule being merged.

> How does this compare to the flake8-self rules?

`private-member-access` doesn't check imports. For example,

```python
from foo import _bar

bar = _bar(...)
```

won't trigger the rule, but would trigger `private-import-name`.

**Comment by charliermarsh** [comment]
Thanks @tjkuson. I assume they would both trigger on:

```python
from foo import _bar

bar = _bar._baz(...)
```

And that only flake8-self would trigger on:

```python
from foo import bar

bar = bar._baz(...)
```

Is that right?

**Comment by tjkuson** [comment]
Yes! Are you thinking of merging them as one rule?

**Comment by charliermarsh** [comment]
@tjkuson - Eventually maybe... For now, I'm okay with them being separate as long as they don't have overlap.

**Comment by tjkuson** [comment]
Rebased to `main`, everything seems to be working. Checking if an import is used as a type annotation will take longer as I need to make it a deferred rule…

**Comment by tjkuson** [comment]
@charliermarsh The PR has been rebased and now forgives private name imports that are used for type annotations!

**Comment by charliermarsh** [comment]
Sorry, this is blocked on me, I've just been slow to get to it.

**Review by charliermarsh — APPROVED** [review]

**Comment by charliermarsh** [comment]
Great work @tjkuson.

**Comment by MichaReiser** [comment]
I'm going to revert this change. It's panicking in the pydantic benchmark run (see https://github.com/astral-sh/ruff/actions/runs/7537305800/job/20516059364?pr=5920).

File: crates/ruff_linter/src/rules/pylint/rules/import_private_name.rs
Comment (MichaReiser) [inline crates/ruff_linter/src/rules/pylint/rules/import_private_name.rs:137]: This doesn't seem to be safe. How is it guaranteed that `index` belongs to `module_name` and not the member name?

Comment (charliermarsh) [reply crates/ruff_linter/src/rules/pylint/rules/import_private_name.rs:137]: Just my fault from a last-minute refactor prior to merging.

Comment (tjkuson) [reply crates/ruff_linter/src/rules/pylint/rules/import_private_name.rs:137]: #9553 slices over `call_path` instead.

**Pull request merged**
